File: java/plugins/com/voxgig/sekreto/plugins/Aws.java
// AWS Secrets Manager and SSM Parameter Store, as voxgig/plugin
// definitions: two kinds, one plugin, because they share a signer.
//
// A PLUGIN, NOT PART OF THE CORE: it opens a socket and signs every
// request. SigV4 travels with it - Sigv4.java is in this folder and not
// in the core, because the core of no port imports a hash function.
// See docs/design/plugin-providers.md.

package com.voxgig.sekreto.plugins;

import static com.voxgig.sekreto.Addr.checkaddr;
import static com.voxgig.sekreto.plugins.Httpjson.dig;
import static com.voxgig.sekreto.plugins.Httpjson.fetchjson;
import static com.voxgig.sekreto.plugins.Httpjson.first;
import static com.voxgig.sekreto.plugins.Httpjson.trimslash;

import com.voxgig.sekreto.Json;
import com.voxgig.sekreto.Provider;
import com.voxgig.sekreto.Sekreto;
import com.voxgig.sekreto.Sekreto.SekretoError;
import com.voxgig.sekreto.Support;
import com.voxgig.sekreto.plugins.Httpjson.Answer;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;
import voxgig.plugin.Definition;

public final class Aws {

  private Aws() {}

  /** The `awssecrets` kind: what `plugins` hands to Sekreto. */
  public static final Definition SECRETS = Support.providerplugin("awssecrets", spec ->
      new Secrets(
          Support.text(spec.get("region")), Support.text(spec.get("keyid")),
          Support.text(spec.get("secret")), Support.text(spec.get("session")),
          Support.text(spec.get("addr"))));

  /** The `awsparams` kind: what `plugins` hands to Sekreto. */
  public static final Definition PARAMS = Support.providerplugin("awsparams", spec ->
      new Params(
          Support.text(spec.get("region")), Support.text(spec.get("keyid")),
          Support.text(spec.get("secret")), Support.text(spec.get("session")),
          Support.text(spec.get("addr")), Support.text(spec.get("prefix"))));

  /** The `YYYYMMDDTHHMMSSZ` timestamp SigV4 wants, for now. */
  static String awsnow() {
    return DateTimeFormatter.ofPattern("yyyyMMdd'T'HHmmss'Z'")
        .withZone(ZoneOffset.UTC)
        .format(Instant.now());
  }

  /** Region and credentials, resolved for one call. */
  private static final class Awsauth {
    final String region;
    final String keyid;
    final String secret;
    final String session;

    Awsauth(String region, String keyid, String secret, String session) {
      this.region = region;
      this.keyid = keyid;
      this.secret = secret;
      this.session = session;
    }
  }

  /**
   * Region and credentials, from config first and the standard AWS_*
   * environment variables second - those are AWS's own convention, and a
   * pod or CI job that has them set should just work. Missing either is an
   * error: an AWS store with no credentials could not answer.
   */
  static Awsauth awsauth(String region, String keyid, String secret, String session) {
    String useregion =
        first(region, System.getenv("AWS_REGION"), System.getenv("AWS_DEFAULT_REGION"));
    String usekeyid = first(keyid, System.getenv("AWS_ACCESS_KEY_ID"));
    String usesecret = first(secret, System.getenv("AWS_SECRET_ACCESS_KEY"));
    String usesession = first(session, System.getenv("AWS_SESSION_TOKEN"));

    if (useregion.isEmpty()) {
      throw new SekretoError("sekreto: aws: no region (set region or AWS_REGION)");
    }
    if (usekeyid.isEmpty() || usesecret.isEmpty()) {
      throw new SekretoError(
          "sekreto: aws: no credentials"
              + " (set keyid/secret or AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY)");
    }

    return new Awsauth(useregion, usekeyid, usesecret,
        usesession.isEmpty() ? null : usesession);
  }

  /** One signed call to an AWS JSON-1.1 API. */
  static Answer awscall(String region, String keyid, String secret, String session,
      String addr, String service, String target, String payload) {
    Awsauth auth = awsauth(region, keyid, secret, session);
    // The China partition lives under its own suffix; every other
    // commercial region is plain amazonaws.com.
    String suffix = auth.region.startsWith("cn-") ? ".amazonaws.com.cn" : ".amazonaws.com";
    String useaddr = first(addr, "https://" + service + "." + auth.region + suffix);
    checkaddr(useaddr);

    String url = trimslash(useaddr) + "/";

    Map<String, Object> extras = new LinkedHashMap<>();
    extras.put("content-type", "application/x-amz-json-1.1");
    extras.put("x-amz-target", target);

    Map<String, Object> input = new LinkedHashMap<>();
    input.put("method", "POST");
    input.put("url", url);
    input.put("headers", extras);
    input.put("body", payload);
    input.put("service", service);
    input.put("region", auth.region);
    input.put("keyid", auth.keyid);
    input.put("secret", auth.secret);
    if (null != auth.session) {
      input.put("session", auth.session);
    }
    input.put("datetime", awsnow());

    Map<String, Object> signed = Sigv4.sigv4(input);

    Map<String, String> send = new LinkedHashMap<>();
    for (Map.Entry<String, Object> entry : extras.entrySet()) {
      send.put(entry.getKey(), String.valueOf(entry.getValue()));
    }
    for (Map.Entry<String, Object> entry : signed.entrySet()) {
      send.put(entry.getKey(), String.valueOf(entry.getValue()));
    }

    return fetchjson("POST", url, send, payload);
  }

  /**
   * Does this AWS error body name one of the not-found types? Those are a
   * miss; every other failure is a store that could not answer.
   */
  static boolean awsmiss(Object body, String... types) {
    Object errtype = dig(body, "__type");
    if (!(errtype instanceof String)) {
      return false;
    }
    for (String type : types) {
      if (((String) errtype).contains(type)) {
        return true;
      }
    }
    return false;
  }

  /**
   * AWS Secrets Manager.
   *
   * <p>`api.token` reads the secret named `api` (the vaultref path, so
   * `db.pass.main` reads `db/pass`) and takes the `token` field of its JSON
   * SecretString - the AWS idiom of one JSON map per secret. A SecretString
   * that is not JSON is the value itself, under the conventional field
   * `value`. Requests are SigV4-signed in-tree; see Sigv4.java.
   */
  public static final class Secrets implements Provider {
    private final String region;
    private final String keyid;
    private final String secret;
    private final String session;
    private final String addr;

    public Secrets(String region, String keyid, String secret, String session, String addr) {
      this.region = region;
      this.keyid = keyid;
      this.secret = secret;
      this.session = session;
      this.addr = addr;
    }

    @Override
    @SuppressWarnings("unchecked")
    public String lookup(String name) {
      Map<String, Object> ref = Sekreto.vaultref(name);

      Map<String, Object> payload = new LinkedHashMap<>();
      payload.put("SecretId", ref.get("path"));

      Answer res = awscall(region, keyid, secret, session, addr,
          "secretsmanager", "secretsmanager.GetSecretValue", Json.stringify(payload));

      if (400 == res.status && awsmiss(res.body, "ResourceNotFoundException")) {
        return null;
      }

      if (200 != res.status) {
        throw new SekretoError("sekreto: aws secretsmanager error: " + res.status);
      }

      Object text = dig(res.body, "SecretString");

      if (!(text instanceof String)) {
        // A binary secret has no fields to address; only the conventional
        // `value` field can mean "the bytes themselves".
        Object bin = dig(res.body, "SecretBinary");
        if (bin instanceof String && "value".equals(ref.get("field"))) {
          // decode() throws IllegalArgumentException on a bad payload,
          // which is not a SekretoError and so escaped the library's own
          // error type. A store that answered incoherently is an error.
          try {
            return new String(
                Base64.getDecoder().decode((String) bin), StandardCharsets.UTF_8);
          } catch (IllegalArgumentException err) {
            throw new SekretoError("sekreto: aws secretsmanager: undecodable secret");
          }
        }
        return null;
      }

      Object parsed = Json.parse((String) text);

      if (parsed instanceof Map) {
        Object value = ((Map<String, Object>) parsed).get(ref.get("field"));
        return null == value ? null : String.valueOf(value);
      }

      // A plain-string secret is the whole value; it has no named fields.
      return "value".equals(ref.get("field")) ? (String) text : null;
    }

    // Config only, never the environment: describe() feeds the spec's
    // sources group, which must answer the same everywhere.
    @Override
    public String describe() {
      return "awssecrets:" + (null == region ? "" : region);
    }
  }

  /**
   * AWS SSM Parameter Store.
   *
   * <p>`db.pass.main` reads the parameter `/db/pass/main` (under an
   * optional prefix path), decrypted. Parameter Store carries flat strings,
   * so there is no field indirection.
   */
  public static final class Params implements Provider {
    private final String region;
    private final String keyid;
    private final String secret;
    private final String session;
    private final String addr;
    private final String prefix;

    public Params(String region, String keyid, String secret, String session,
        String addr, String prefix) {
      this.region = region;
      this.keyid = keyid;
      this.secret = secret;
      this.session = session;
      this.addr = addr;
      this.prefix = prefix;
    }

    @Override
    public String lookup(String name) {
      Map<String, Object> payload = new LinkedHashMap<>();
      payload.put("Name", Sekreto.awsparam(name, prefix));
      payload.put("WithDecryption", Boolean.TRUE);

      Answer res = awscall(region, keyid, secret, session, addr,
          "ssm", "AmazonSSM.GetParameter", Json.stringify(payload));

      if (400 == res.status && awsmiss(res.body, "ParameterNotFound")) {
        return null;
      }

      if (200 != res.status) {
        throw new SekretoError("sekreto: aws ssm error: " + res.status);
      }

      Object value = dig(res.body, "Parameter", "Value");
      return null == value ? null : String.valueOf(value);
    }

    @Override
    public String describe() {
      return "awsparams:" + (null == region ? "" : region) + (null == prefix ? "" : prefix);
    }
  }
}

File: clojure/plugins/voxgig/sekreto/plugins/aws.clj
;; The two AWS stores, as voxgig/plugin definitions - and the only two
;; kinds that sign a request.
;;
;; ONE PLUGIN, TWO KINDS: Secrets Manager and Parameter Store share a
;; signer, so they ship together and `sigv4` comes with them. That is what
;; keeps the core of every port free of a hash function
;; (docs/design/plugin-providers.md).
;;
;; A port of typescript/plugins/aws.ts, which is canonical.

(ns voxgig.sekreto.plugins.aws
  (:require [clojure.string :as string]
            [voxgig.sekreto.core :as core]
            [voxgig.sekreto.addr :refer [checkaddr]]
            [voxgig.sekreto.json :as json]
            [voxgig.sekreto.provider :as provider]
            [voxgig.sekreto.providers :as providers]
            [voxgig.sekreto.plugins.httpjson :as http]
            [voxgig.sekreto.plugins.sigv4 :as sigv4])
  (:import [java.nio.charset StandardCharsets]
           [java.time Instant ZoneOffset]
           [java.time.format DateTimeFormatter]
           [java.util Base64]))

(defn awsnow
  "The `YYYYMMDDTHHMMSSZ` timestamp SigV4 wants, for now."
  []
  (.format (.withZone (DateTimeFormatter/ofPattern "yyyyMMdd'T'HHmmss'Z'") ZoneOffset/UTC)
           (Instant/now)))

(defn awsauth
  "Region and credentials, from config first and the standard AWS_*
  environment variables second - those are AWS's own convention, and a pod
  or CI job that has them set should just work. Missing either is an error:
  an AWS store with no credentials could not answer."
  [{:keys [region keyid secret session]}]
  (let [useregion (http/firstof region (core/getenv "AWS_REGION") (core/getenv "AWS_DEFAULT_REGION"))
        usekeyid (http/firstof keyid (core/getenv "AWS_ACCESS_KEY_ID"))
        usesecret (http/firstof secret (core/getenv "AWS_SECRET_ACCESS_KEY"))
        usesession (http/firstof session (core/getenv "AWS_SESSION_TOKEN"))]

    (when (= "" useregion)
      (throw (core/sekretoerror "sekreto: aws: no region (set region or AWS_REGION)")))

    (when (or (= "" usekeyid) (= "" usesecret))
      (throw (core/sekretoerror
              (str "sekreto: aws: no credentials"
                   " (set keyid/secret or AWS_ACCESS_KEY_ID/AWS_SECRET_ACCESS_KEY)"))))

    {:region useregion :keyid usekeyid :secret usesecret :session (core/notempty usesession)}))

(defn awscall
  "One signed call to an AWS JSON-1.1 API."
  [config service target payload]
  (let [auth (awsauth config)
        ;; The China partition lives under its own suffix; every other
        ;; commercial region is plain amazonaws.com.
        suffix (if (string/starts-with? (:region auth) "cn-") ".amazonaws.com.cn" ".amazonaws.com")
        useaddr (http/firstof (:addr config) (str "https://" service "." (:region auth) suffix))
        _ (checkaddr useaddr)
        url (str (http/trimslash useaddr) "/")
        extras (json/omap [["content-type" "application/x-amz-json-1.1"] ["x-amz-target" target]])
        signed (sigv4/sigv4 {:method "POST"
                             :url url
                             :service service
                             :region (:region auth)
                             :keyid (:keyid auth)
                             :secret (:secret auth)
                             :datetime (awsnow)
                             :headers extras
                             :body payload
                             :session (:session auth)})]
    (http/fetchjson "POST" url (merge extras signed) payload)))

(defn awsmiss?
  "Does this AWS error body name one of the not-found types? Those are a
  miss; every other failure is a store that could not answer."
  [body & types]
  (let [errtype (json/asstr (json/dig body "__type"))]
    (boolean (and errtype (some (fn [want] (string/includes? errtype want)) types)))))

(defrecord Awssecrets [region keyid secret session addr]
  provider/Provider

  ;; AWS Secrets Manager.
  ;;
  ;; `api.token` reads the secret named `api` (the vaultref path, so
  ;; `db.pass.main` reads `db/pass`) and takes the `token` field of its JSON
  ;; SecretString - the AWS idiom of one JSON map per secret. A SecretString
  ;; that is not JSON is the value itself, under the conventional field
  ;; `value`. Requests are SigV4-signed in-tree; see sigv4.clj.
  (lookup [this name]
    (let [ref (core/vaultref name)
          res (awscall this "secretsmanager" "secretsmanager.GetSecretValue"
                       (json/stringify (json/omap [["SecretId" (:path ref)]])))]

      (cond
        (and (= 400 (:status res)) (awsmiss? (:body res) "ResourceNotFoundException")) nil

        (not= 200 (:status res))
        (throw (core/sekretoerror (str "sekreto: aws secretsmanager error: " (:status res))))

        :else
        (if-let [text (json/asstr (json/dig (:body res) "SecretString"))]
          (let [fields (json/parse text)]
            (if (map? fields)
              (json/astext (json/dig fields (:field ref)))
              ;; A plain-string secret is the whole value; it has no named
              ;; fields.
              (when (= "value" (:field ref)) text)))

          ;; A binary secret has no fields to address; only the conventional
          ;; `value` field can mean "the bytes themselves".
          (let [bin (json/asstr (json/dig (:body res) "SecretBinary"))]
            (when (and bin (= "value" (:field ref)))
              (try
                (String. (.decode (Base64/getDecoder) ^String bin) StandardCharsets/UTF_8)
                ;; A store that answered incoherently is an error, and the
                ;; decoder's own complaint is not one of this library's.
                (catch IllegalArgumentException _
                  (throw (core/sekretoerror
                          "sekreto: aws secretsmanager: undecodable secret"))))))))))

  ;; Config only, never the environment: `describe` feeds the spec's sources
  ;; group, which must answer the same everywhere.
  (describe [_] (str "awssecrets:" (or region ""))))

(defrecord Awsparams [region keyid secret session addr prefix]
  provider/Provider

  ;; AWS SSM Parameter Store.
  ;;
  ;; `db.pass.main` reads the parameter `/db/pass/main` (under an optional
  ;; prefix path), decrypted. Parameter Store carries flat strings, so there
  ;; is no field indirection.
  (lookup [this name]
    (let [payload (json/omap [["Name" (core/awsparam name prefix)] ["WithDecryption" true]])
          res (awscall this "ssm" "AmazonSSM.GetParameter" (json/stringify payload))]

      (cond
        (and (= 400 (:status res)) (awsmiss? (:body res) "ParameterNotFound")) nil

        (not= 200 (:status res))
        (throw (core/sekretoerror (str "sekreto: aws ssm error: " (:status res))))

        :else (json/astext (json/dig (:body res) "Parameter" "Value")))))

  (describe [_] (str "awsparams:" (or region "") (or prefix ""))))

(def awssecrets
  "The `awssecrets` provider kind: AWS Secrets Manager."
  (providers/providerplugin
   "awssecrets"
   (fn [{:keys [region keyid secret session addr]}]
     (->Awssecrets region keyid secret session addr))))

(def awsparams
  "The `awsparams` provider kind: AWS SSM Parameter Store."
  (providers/providerplugin
   "awsparams"
   (fn [{:keys [region keyid secret session addr prefix]}]
     (->Awsparams region keyid secret session addr prefix))))
